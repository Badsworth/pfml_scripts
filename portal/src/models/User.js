import BaseModel from "./BaseModel";

class User extends BaseModel {
  get defaults() {
    return {
      user_id: null,
      auth_id: null,
      email_address: null,
      first_name: null,
      middle_name: null,
      last_name: null,
      date_of_birth: null,
      has_state_id: null,
      state_id: null,
      status: null,
      ssn_or_itin: null,
    };
  }
}

export default User;
