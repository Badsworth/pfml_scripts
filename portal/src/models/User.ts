/* eslint sort-keys: ["error", "asc"] */
export const PhoneType = {
  cell: "Cell",
  phone: "Phone",
} as const;

export const MFAPreference = {
  optOut: "Opt Out",
  sms: "SMS",
} as const;

class User {
  auth_id: string;
  consented_to_data_sharing: boolean;
  email_address: string;
  roles: UserRole[] = [];
  user_id: string;
  mfa_phone_number: {
    int_code: string | null;
    phone_number: string | null;
    phone_type: typeof PhoneType[keyof typeof PhoneType] | null;
  } | null = null;

  mfa_delivery_preference:
    | typeof MFAPreference[keyof typeof MFAPreference]
    | null = null;

  user_leave_administrators: UserLeaveAdministrator[] = [];

  constructor(attrs: Partial<User>) {
    Object.assign(this, attrs);
  }

  get hasEmployerRole() {
    return this.roles.some(
      (userRole) => userRole.role_description === RoleDescription.employer
    );
  }

  /**
   * Determines whether user_leave_administrators has only unverified employers
   */
  get hasOnlyUnverifiedEmployers(): boolean {
    return this.verifiedEmployers.length === 0;
  }

  /**
   * Determines whether user_leave_administrators has a verifiable employer
   */
  get hasVerifiableEmployer(): boolean {
    return this.user_leave_administrators.some((employer) =>
      this.isVerifiableEmployer(employer)
    );
  }

  /**
   * Returns list of verified employers
   */
  get verifiedEmployers(): UserLeaveAdministrator[] {
    return this.user_leave_administrators.filter(
      (employer) => employer.verified === true
    );
  }

  /**
   * Determines whether an employer is verifiable (unverified and has verification data)
   */
  isVerifiableEmployer(employer: UserLeaveAdministrator): boolean {
    return !employer.verified && employer.has_verification_data;
  }

  /**
   * Determines whether an employer is registered in FINEOS
   */
  isEmployerIdRegisteredInFineos(employerId: string | null): boolean {
    return this.user_leave_administrators.some(
      (employer) =>
        employerId === employer.employer_id && employer.has_fineos_registration
    );
  }
}

export class UserRole {
  role_description: typeof RoleDescription[keyof typeof RoleDescription];
  role_id: number;

  constructor(attrs: Partial<UserRole>) {
    Object.assign(this, attrs);
  }
}

/**
 * Enums for UserRole's `role_description` field
 * @enum {string}
 */
export const RoleDescription = {
  claimant: "Claimant",
  employer: "Employer",
} as const;

export class UserLeaveAdministrator {
  employer_dba: string;
  employer_fein: string;
  employer_id: string;
  has_fineos_registration: boolean;
  has_verification_data: boolean;
  verified: boolean;

  constructor(attrs: Partial<UserLeaveAdministrator>) {
    Object.assign(this, attrs);
  }
}

export default User;
