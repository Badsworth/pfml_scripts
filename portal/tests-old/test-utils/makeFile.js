/**
 * Make a File instance
 * @see https://developer.mozilla.org/en-US/docs/Web/API/File
 * @param {object} attrs File attributes
 * @param {string} attrs.name The file's name
 * @param {string} attrs.type The file's MIME type
 * @returns {File}
 */
export const makeFile = (attrs = {}) => {
  const { name, type } = Object.assign(
    {
      name: "file.pdf",
      type: "application/pdf",
    },
    attrs
  );

  return new File([], name, { type });
};

/**
 * Create an HTML <input> for a test's event.target
 * @param {object} attrs - HTML attributes for the input
 * @returns {HTMLInputElement}
 */
export function createInputElement(attrs) {
  const input = document.createElement("input");

  Object.entries(attrs).forEach(([attributeName, value]) => {
    input.setAttribute(attributeName, value);
  });

  if (attrs.checked === false) input.checked = false;

  return input;
}
