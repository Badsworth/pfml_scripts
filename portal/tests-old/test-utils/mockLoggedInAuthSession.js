import { Auth } from "@aws-amplify/auth";

/**
 * Mock the Cognito authentication session so a test user appears logged in
 */
const mockLoggedInAuthSession = () => {
  const accessTokenJwt =
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiQnVkIn0.YDRecdsqG_plEwM0H8rK7t2z0R3XRNESJB5ZXk-FRN8";

  jest.spyOn(Auth, "currentSession").mockImplementation(() =>
    Promise.resolve({
      accessToken: { jwtToken: accessTokenJwt },
    })
  );
};

export default mockLoggedInAuthSession;
