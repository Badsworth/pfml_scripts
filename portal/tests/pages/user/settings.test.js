import User, { RoleDescription } from "../../../src/models/User";
import { act, screen } from "@testing-library/react";
import Settings from "../../../src/pages/user/settings";
import { renderPage } from "../../test-utils";
import userEvent from "@testing-library/user-event";

jest.mock("../../../src/services/tracker");

beforeEach(() => {
  process.env.featureFlags = JSON.stringify({ claimantShowMFA: true });
});

const updateUser = jest.fn();
const setMFAPreference = jest.fn();

const renderWithUser = (
  user = new User({
    email_address: "mock@gmail.com",
    consented_to_data_sharing: true,
  })
) => {
  const props = {
    user,
    query: {},
  };
  return renderPage(
    Settings,
    {
      addCustomSetup: (appLogic) => {
        appLogic.users.user = user;
        appLogic.users.updateUser = updateUser;
        appLogic.auth.setMFAPreference = setMFAPreference;
      },
    },
    props
  );
};

describe(Settings, () => {
  it("renders the page", () => {
    const { container } = renderWithUser();
    expect(
      screen.getByText(/Additional login verification is not enabled/)
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  describe("when MFA is enabled", () => {
    const user = new User({
      user_id: "mock-user-id",
      mfa_delivery_preference: "SMS",
      mfa_phone_number: {
        int_code: "1",
        phone_number: "***-***-1234",
        phone_type: "Cell",
      },
      email_address: "mock@gmail.com",
      consented_to_data_sharing: true,
    });

    it("shows MFA configuration", () => {
      const { container } = renderWithUser(user);
      expect(
        screen.getByText(/Additional login verification is enabled/)
      ).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    it("shows amendment form when user edits their login verification", async () => {
      renderWithUser(user);

      const editButton = screen.getAllByRole("button", { name: /Edit/ })[0];

      await act(async () => await userEvent.click(editButton));

      expect(
        await screen.findByText(/Edit login verification preferences/)
      ).toBeInTheDocument();
    });

    describe("and user is editing login verification", () => {
      beforeEach(async () => {
        renderWithUser(user);

        const editButton = screen.getAllByRole("button", { name: /Edit/ })[0];

        await act(async () => await userEvent.click(editButton));
      });

      it("hides amendment when user clicks cancel", async () => {
        const cancelButton = screen.getByRole("button", { name: /Cancel/ });
        await act(async () => await userEvent.click(cancelButton));

        expect(
          screen.queryByText(/Edit login verification preferences/)
        ).not.toBeInTheDocument();
      });

      it("changes radio value when user clicks an option", async () => {
        const disableRadio = screen.getByText(
          /Disable additional login verification/
        );
        await act(async () => await userEvent.click(disableRadio));
        expect(
          screen.getByLabelText(/Disable additional login verification/)
        ).toBeChecked();
        expect(
          screen.getByLabelText(/Enable additional login verification/)
        ).not.toBeChecked();

        const enableRadio = screen.getByText(
          /Enable additional login verification/
        );
        await act(async () => await userEvent.click(enableRadio));
        expect(
          screen.getByLabelText(/Disable additional login verification/)
        ).not.toBeChecked();
        expect(
          screen.getByLabelText(/Enable additional login verification/)
        ).toBeChecked();
      });

      it("saves mfa preference when user clicks save", async () => {
        const saveButton = screen.getByRole("button", {
          name: /Save preference/,
        });
        await act(async () => await userEvent.click(saveButton));

        expect(updateUser).toHaveBeenCalledWith(expect.any(String), {
          mfa_delivery_preference: "SMS",
        });
        expect(
          screen.queryByText(/Edit login verification preferences/)
        ).not.toBeInTheDocument();
      });
    });
  });

  it("displays success alert when user sets up SMS MFA", () => {
    renderPage(Settings, {}, { query: { smsMfaConfirmed: "true" } });
    expect(screen.getByRole("region")).toMatchSnapshot();
  });

  it("renders 404 when claimantShowMFA feature flag is disabled", () => {
    process.env.featureFlags = JSON.stringify({ claimantShowMFA: false });
    renderPage(Settings);

    const pageNotFoundHeading = screen.getByRole("heading", {
      name: /Page not found/,
    });
    expect(pageNotFoundHeading).toBeInTheDocument();
  });

  it("renders 404 when user has employer role", () => {
    const user = new User({
      user_id: "mock-user-id",
      mfa_delivery_preference: "SMS",
      mfa_phone_number: {
        int_code: "1",
        phone_number: "***-***-1234",
        phone_type: "Cell",
      },
      email_address: "mock@gmail.com",
      consented_to_data_sharing: true,
      roles: [{ role_description: RoleDescription.employer, role_id: 1 }],
    });
    renderWithUser(user);
    const pageNotFoundHeading = screen.getByRole("heading", {
      name: /Page not found/,
    });
    expect(pageNotFoundHeading).toBeInTheDocument();
  });
});
