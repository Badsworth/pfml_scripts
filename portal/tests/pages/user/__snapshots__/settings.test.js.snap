// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentWithUser renders the page 1`] = `
<div>
  <div
    class="measure-6"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-6 font-heading-lg line-height-sans-2"
      tabindex="-1"
    >
      Settings
    </h1>
    <div
      class="display-flex flex-align-end margin-top-6 margin-bottom-3"
    >
      <h3
        class="flex-fill margin-right-1 font-heading-md text-bold"
      >
        Account information
      </h3>
    </div>
    <div
      class="margin-bottom-2 padding-bottom-2 display-flex flex-justify"
    >
      <div
        class="margin-right-2"
      >
        <h3
          class="margin-bottom-1 font-heading-xs text-bold"
        >
          Email address
        </h3>
        mock@gmail.com
      </div>
    </div>
    <div
      class="display-flex flex-align-end margin-top-6 margin-bottom-3"
    >
      <h3
        class="flex-fill margin-right-1 font-heading-md text-bold"
      >
        Additional login verification
      </h3>
    </div>
    <p>
      Make your account more secure by providing a phone number to associate with your account. We'll send you a 6-digit code whenenever we need to verify it's really you. You’ll need to enter the code in order to log in to your account. In the future, you’ll also be able to access sensitive information like tax documents once you set this up.
    </p>
    <p>
      You will need to provide a phone number that can receive text messages (SMS) and that you’ll have consistent access to in case we need to verify your login.
    </p>
    <p>
      <strong>
        Additional login verification is not enabled
      </strong>
    </p>
    <a
      class="usa-button"
      href="/applications"
    >
      Add phone number
    </a>
  </div>
</div>
`;

exports[`ComponentWithUser when MFA is enabled shows MFA configuration 1`] = `
<div>
  <div
    class="measure-6"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-6 font-heading-lg line-height-sans-2"
      tabindex="-1"
    >
      Settings
    </h1>
    <div
      class="display-flex flex-align-end margin-top-6 margin-bottom-3"
    >
      <h3
        class="flex-fill margin-right-1 font-heading-md text-bold"
      >
        Account information
      </h3>
    </div>
    <div
      class="margin-bottom-2 padding-bottom-2 display-flex flex-justify"
    >
      <div
        class="margin-right-2"
      >
        <h3
          class="margin-bottom-1 font-heading-xs text-bold"
        >
          Email address
        </h3>
        mock@gmail.com
      </div>
    </div>
    <div
      class="display-flex flex-align-end margin-top-6 margin-bottom-3"
    >
      <h3
        class="flex-fill margin-right-1 font-heading-md text-bold"
      >
        Additional login verification
      </h3>
    </div>
    <p>
      To protect your security, we'll send you a 6-digit code whenenever we need to verify it's really you. You’ll need to enter the code in order to log in to your account.
    </p>
    <div
      class="margin-bottom-2 padding-bottom-2 display-flex flex-justify border-bottom-2px border-base-lighter"
    >
      <div
        class="margin-right-2"
      >
        <h3
          class="margin-bottom-1 font-heading-xs text-bold"
        >
          Additional login verification is enabled
        </h3>
      </div>
      <button
        class="usa-button position-relative width-auto display-flex usa-button--unstyled"
        type="button"
      >
        <strong>
          Edit
        </strong>
      </button>
    </div>
    <div
      class="margin-bottom-2 padding-bottom-2 display-flex flex-justify border-bottom-2px border-base-lighter"
    >
      <div
        class="margin-right-2"
      >
        <h3
          class="margin-bottom-1 font-heading-xs text-bold"
        >
          Phone number
        </h3>
        ***-***-1234
      </div>
      <a
        aria-label="Edit: Phone number"
        class="usa-link"
        href="/applications"
      >
        Edit
      </a>
    </div>
  </div>
</div>
`;
